File: Programmers/python/게임 맵 최단거리.py
from collections import deque

def bfs(maps, cur_x, cur_y):
    min_dist = 10000000
    dy = [0, -1, 0, 1]
    dx = [-1, 0, 1, 0]
    end_y = len(maps) - 1
    end_x = len(maps[0]) - 1
    visited = [[False for _ in range(len(maps[0]))] for _ in range(len(maps))]
    dq = deque([[(cur_y, cur_x), 1]])
    
    visited[cur_y][cur_x] = True
    while dq:
        (cur_y, cur_x), dist = dq.popleft()
        
        for i in range(4):
            # print(dq)
            next_y = cur_y + dy[i]
            next_x = cur_x + dx[i]
            
            if next_x < 0 or next_y < 0 or next_x > end_x or next_y > end_y:
                continue
            if maps[next_y][next_x] == 1 and not visited[next_y][next_x]:
                dq.append([(next_y, next_x), dist + 1])
                visited[next_y][next_x] = True
                if next_y == end_y and next_x == end_x:
                    min_dist = min(min_dist, dist + 1)
    return min_dist


def solution(maps):
    dist = 0
    dist = bfs(maps, 0, 0)
    
    if dist == 10000000:
        return -1
    
    return dist

File: Programmers/java/게임 맵 최단거리.java
import java.util.*;

class Solution {
    public int solution(int[][] maps) {
        return bfs(maps);
    }
    
    public int bfs(int[][] maps) {
        int[] dy = {0, -1, 0, 1}; // left, up, right, down
        int[] dx = {-1, 0, 1, 0};
        int maxHeight = maps.length;
        int maxWidth = maps[0].length;
        
        ArrayList<ArrayList<Integer>> graph = new ArrayList<ArrayList<Integer>>();
        boolean[][] isVisited = new boolean[maxHeight][maxWidth];
        Queue<Pos> queue = new LinkedList<>();
        
        queue.add(new Pos(0, 0, 1));
        isVisited[0][0] = true;
        
        while(!queue.isEmpty()) {
            Pos cur = queue.poll();
            
            if(cur.y == maxHeight - 1 && cur.x == maxWidth - 1) {
                return cur.dist;
            }
            
            for(int i = 0; i < 4; i++){
                int nextY = cur.y + dy[i];
                int nextX = cur.x + dx[i];
                
                if(isValidPos(maxHeight, maxWidth, nextY, nextX)) {
                    // 유효한 좌표일 때에만
                    if(!isVisited[nextY][nextX] && maps[nextY][nextX] == 1) {
                        isVisited[nextY][nextX] = true;
                        queue.add(new Pos(nextY, nextX, cur.dist + 1));
                    }
                }
            }
        }
        
        return -1;
    }
    
    boolean isValidPos(int maxHeight, int maxWidth, int curY, int curX) {
        if (curY >= maxHeight || curX >= maxWidth) {
            return false;
        }
        if(curY < 0 || curX < 0){
            return false;
        }
        return true;
    }
    
    class Pos{
        int y;
        int x;
        int dist;
        
        public Pos(int y, int x, int dist) {
            this.y = y;
            this.x = x;
            this.dist = dist;
        }
    }
}
